meetings/working-groups/extensions/content-based-naming.md

```diff
@@ -0,0 +1,217 @@
+    2. Changing C# specific aspects of the type like adding / removing nullable annotations, changing tuple names, etc ...
+3. The C# compiler should be able to fully rehydrate the original extension block from metadata. This means for each skeleton member the compiler can recreate the _exact_ type name that was written in source. This includes items like the original type parameter names, nullable annotations, etc ...
+
+This will be achieved by moving to a content based naming scheme for skeleton types and members.
+
+## Detailed Design
+
+This proposal relies heavily on using content based naming for skeleton types and members.  The proposal is going to focus on what items are included in the content name, it will not discuss the actual name produced. That is because the specific hashing algorithm used will be an implementation detail of the compiler.
+
+For an extension block the content name of the skeleton type will be determined by using the following parts of the extension type declaration:
+
+- The fully qualified CLR name of the type. This will include namespaces, type constraints, named constraints (like `new`). Constraints will be ordered in the content name such that reordering them in source code does not change the name. Specifically:
+    - Type parameter constraints will be listed in declaration order. The constraints for the Nth type parameter will occur before the Nth+1 type parameter.
+    - Base type and interface constraints will be sorted by comparing the full names ordinally
+    - Non-type constraints will be sorted by comparing the C# text ordinally
+- This will not include any C# isms like tuple names, nullability, etc ... 
+
+The type parameters of the skeleton type will match those of the extension block. The exception being that the name of the type parameters will be normalized to `T0`, `T1`, etc ... based on the order they appear in the type declaration. This means an extension block with a type parameter `Dictionary<TKey, TValue>` will result in a skeleton type with type parameters `T0` and `T1` in that order.
+
+This will achieve the goal that the name of skeleton types will only change when the underlying CLR type of an extension type changes. Any change to C# parts of the type such as adding nullable annotations or tuple names will impact the skeleton type name.
+
+For an extension block a `private` method, referred to as the source type method, will be generated in the skeleton type. This method will have a single parameter which mirrors the extension `this` parameter. Specifically it will have the same attributes, ref declaration, parameter name and C# type. The name of this method will be a content name that is determined by using the following parts of the extension parameter declaration:
+
+- The fully qualified C# name of the type. This will include items like nullable annotations, tuple names, etc ... The constraints will have the same ordering as CLR types for the skeleton type name.
```

Comment:
> The constraints will have the same ordering as CLR types for the skeleton type name.

This is probably better to have as a separate bullet point, explaining in more details since this method is not generic and it is not clear what constraints we are talking about. I assume you are referring to extension block type parameter constraint types, but now using C# type names.